flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerServiceImpl.java

```diff
@@ -307,18 +307,36 @@ void onProcessingTime(long time) throws Exception {
     }
 
     public void advanceWatermark(long time) throws Exception {
-        currentWatermark = time;
+        Preconditions.checkState(
+                tryAdvanceWatermark(
+                        time,
+                        () -> {
+                            // Never stop advancing.
+                            return false;
+                        }));
+    }
 
+    /**
+     * @return true if following watermarks can be processed immediately. False if the firing timers
+     *     should be interrupted as soon as possible.
+     */
+    public boolean tryAdvanceWatermark(
+            long time, InternalTimeServiceManager.ShouldStopAdvancingFn shouldStopAdvancingFn)
+            throws Exception {
+        currentWatermark = time;
         InternalTimer<K, N> timer;
-
         while ((timer = eventTimeTimersQueue.peek()) != null
                 && timer.getTimestamp() <= time
                 && !cancellationContext.isCancelled()) {
             keyContext.setCurrentKey(timer.getKey());
             eventTimeTimersQueue.poll();
             triggerTarget.onEventTime(timer);
             taskIOMetricGroup.getNumFiredTimers().inc();
+            if (shouldStopAdvancingFn.shouldStopAdvancing()) {
+                return false;
+            }
```

Comment:
I've added a comment explaining why this is at the end:
```
            // Check if we should stop advancing after at least one iteration to guarantee progress
            // and prevent a potential starvation.
```

Reply:
If so I would suggest another `boolean first` to suppress the check in first round. It would achieve both aspects. WDYT?

Reply:
Instead of `boolean first` I've added `boolean stop` and re-used it both for cancellation and interruption. I didn't want to make the `while()` condition any longer, as it would become too difficult to read IMO. 

WDYT?